core/imds/src/test/java/software/amazon/awssdk/imds/Ec2MetadataWithApacheClientTest.java

```diff
@@ -0,0 +1,389 @@
+
+    private static final String TOKEN_HEADER = "x-aws-ec2-metadata-token";
+
+    private static final String EC2_METADATA_TOKEN_TTL_HEADER = "x-aws-ec2-metadata-token-ttl-seconds";
+
+    private static final String EC2_METADATA_ROOT = "/latest/meta-data";
+
+    private static final String AMI_ID_RESOURCE = EC2_METADATA_ROOT + "/ami-id";
+
+    private SdkHttpClient httpClient;
+
+    @Mock
+    private Ec2Metadata ec2Metadata;
+
+    @Rule
+    public WireMockRule mockMetadataEndpoint = new WireMockRule();
+
+    @Before
+    public void methodSetup() {
+        System.setProperty(SdkSystemSetting.AWS_EC2_METADATA_SERVICE_ENDPOINT.property(), "http://localhost:" + mockMetadataEndpoint.port());
+        httpClient = ApacheHttpClient.create();
+    }
+
+    @Test
+    public void get_whenDummyResponseIsReturned(){
+
+        MetadataResponse metadataResponse = new MetadataResponse("IMDS");
+        when(ec2Metadata.get("/ami-id")).thenReturn(metadataResponse);
+        assertThat(ec2Metadata.get("/ami-id").asString()).isEqualTo("IMDS");
+
+    }
+
+    @Test
+    public void verifyEc2Metadata_equalsAndHashcode(){
+
+        EqualsVerifier.forClass(DefaultEc2Metadata.class)
+            .usingGetClass()
+            .verify();
+    }
+
+    @Test
+    public void get_shouldSucceedOnFirstAttempt() throws IOException {
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).willReturn(aResponse().withBody("some-token")));
+        stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE)).willReturn(aResponse().withBody("{}")));
+
+        Ec2Metadata ec2Metadata =
+            Ec2Metadata.builder().httpClient(httpClient).endpoint(URI.create("http://localhost:8080")).build();
+
+        MetadataResponse metadataResponse = ec2Metadata.get("/latest/meta-data/ami-id");
+        assertThat(metadataResponse.asString()).isEqualTo("{}");
+
+        WireMock.verify(putRequestedFor(urlPathEqualTo(TOKEN_RESOURCE_PATH)).withHeader(EC2_METADATA_TOKEN_TTL_HEADER, equalTo("21600")));
+        WireMock.verify(getRequestedFor(urlPathEqualTo(AMI_ID_RESOURCE)).withHeader(TOKEN_HEADER, equalTo("some-token")));
+    }
+
+    @Test
+    public void get_failedThriceWith404() throws IOException {
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).willReturn(aResponse().withBody("some-token")));
+        stubFor(get(urlPathEqualTo(AMI_ID_RESOURCE)).willReturn(aResponse().withBody("{}").withStatus(404)));
+
+        assertThatThrownBy(() -> {
+            Ec2Metadata ec2Metadata =
+                Ec2Metadata.builder().httpClient(httpClient).endpoint(URI.create("http://localhost:8080")).build();
+            MetadataResponse metadataResponse = ec2Metadata.get("/latest/meta-data/ami-id");
+        }).hasMessageContaining("metadata")
+          .isInstanceOf(SdkClientException.class);
+    }
+
+    @Test
+    public void get_failedThriceWith401() throws IOException {
+
+        stubFor(put(urlPathEqualTo(TOKEN_RESOURCE_PATH)).willReturn(aResponse().withBody("some-token")));
```

Comment:
Here too are we giving details like "Unauthorized response status code indicates that the client request has not been completed because it lacks valid authentication credentials for the requested resource " for 401 ?

Reply:
According to the SEP, 401 is a Retry-able exception . So no specific message is been given out.